File: 정종길/프로그래머스/카카오 3회독/2023 KAKAO BLIND RECRUITMENT/택배 배달과 수거하기.java
import java.util.*;

class Solution {
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        long answer = 0;
        int dEnd = -1, pEnd = -1;
        
        for(int i = n - 1; i >= 0; i--){
            if(deliveries[i] != 0){
                dEnd = i;
                break;
            }
        }
        
        for(int j = n - 1; j >= 0; j--){
            if(pickups[j] != 0){
                pEnd = j;
                break;
            }
        }

        while(dEnd != -1 || pEnd != -1) {
            answer += (Math.max(dEnd, pEnd) + 1) * 2;
            int dcap = cap;
            
            while(dEnd != -1){
                if(deliveries[dEnd] > dcap){
                    deliveries[dEnd] -= dcap;
                    break;
                    
                } else {
                    dcap -= deliveries[dEnd];
                    deliveries[dEnd] = 0;
                }
                
                while(dEnd != -1 && deliveries[dEnd] == 0){
                    dEnd -= 1;
                }
            }
            
            int pcap = cap;
            while(pEnd != -1){
                if(pickups[pEnd] > pcap){
                    pickups[pEnd] -= pcap;
                    break;

                } else {
                    pcap -= pickups[pEnd];
                    pickups[pEnd] = 0;
                }

                while(pEnd != -1 && pickups[pEnd] == 0){
                    pEnd -= 1;
                }
            }
        }
        
        return answer;
    }
}

File: 정종길/프로그래머스/카카오 3회독/2023 KAKAO BLIND RECRUITMENT/택배 배달과 수거하기.py
'''
1. 배달이던 수거던 어짜피 가장 먼 곳까지는 가야 됨
 -> 배달의 경우는 꽉 채워 출발해서 가는 길에 다 뿌려 주면 됨
 -> 수거의 경우 오는 길에 다 수집해서 오면 됨
2. 가장 먼 곳을 계속 왕복 하면서, 배달 수거를 지워가기
'''

def solution(cap, n, deliveries, pickups):
    answer = 0
    d = p = n - 1
    
    while d >= 0 or p >= 0:
        cnt = cap
        while d >= 0 and not deliveries[d]:
            d -= 1
        while p >= 0 and not pickups[p]:
            p -= 1
    
        answer += max(d + 1, p + 1) * 2
        
        d = act(deliveries, d, cap)
        p = act(pickups, p, cap)
        
    return answer

def act(li, here, cap):
    while here >= 0:
        if cap <= li[here]:
            li[here] -= cap
            return here
        else:
            cap -= li[here]
            li[here] = 0
            here -= 1
    return here